client/src/test/java/io/pravega/client/stream/mock/MockStreamManager.java

```diff
@@ -181,6 +193,30 @@ public boolean deleteStream(String scopeName, String toDelete) {
         return Futures.getAndHandleExceptions(controller.deleteStream(scopeName, toDelete), RuntimeException::new);
     }
 
+    @Override
+    public Iterator<String> listScopes() {
+        AsyncIterator<String> asyncIterator = controller.listScopes();
```

Comment:
Why not use `asyncIterator.asIterator`?

You implementation below is incorrect. Both `hasNext` and `next` will invoke `load` which will cause the wrapped iterator to advance. This means you will be skipping over iterator items as you iterate through. Also `next` is supposed to throw if `hasNext() == false`

Reply:
yes, but the load doesnt call asynciterator.getNext unless it has exhausted all existing entries. 
so wrapper iterator will only advance when its getNext gets called and if it returns items then hasNext returns true, false otherwise.

Reply:
And next throws exception if you call next when hasNext may have written false.

Reply:
The point is that you are returning an `Iterator` object which does not conform to the `Iterator` [contracts](https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html). If you want to implement a different behavior you will have to return something that does not implement `java.util.Iterator`

Reply:
i dont completely follow how the behaviour is different for BlockingAsyncIterator from the iterator contract. 
Can you please illustrate with an example.

It upholds following contracts (does not provide an implementation for `remove` method as that is unsupported for this iterator).

- if there are elements to iterate over, `hasNext` will return true. 
- `next` returns next element in the iteration "without skipping over any element".

Following is the implementation for load method which is called from both hashNext and next which loads next and canHaveNext values into fields `T next` and `boolean canHaveNext`.
then hasNext simply returns canHaveNext and next returns next or throws if canHaveNext is false and next is called.
```
    private void load() {
        if (next == null && canHaveNext) {
            next = asyncIterator.getNext().join();
            if (next == null) {
                canHaveNext = false;
            }
        }
    }
```

`forEachRemaining` would work correctly as it performs `hasNext` followed by `next` call in a loop.

Reply:
Oh. My Bad.

 I didnt realize that this mock class also has an implementation.. yes, that is perhaps incorrect. 
i was talking about BlockingAsyncIterator.